application/src/Controller/MatrixController.php

```diff
@@ -273,6 +273,92 @@ public function createRoom(string $serverID, Request $request):JsonResponse {
         return new JsonResponse((object) $response, 200);
     }
 
+    /**
+     * Get room power levels.
+     *
+     * @Route("/v3/sync")
+     * @param string $serverID
+     * @param Request $request
+     * @return JsonResponse
+     */
+    public function getRoomPowerLevels(string $serverID, Request $request):JsonResponse {
+        // 1. Check call auth.
+        // 2. Check HTTP method is accepted.
+        $accessCheck = $this->authHttpCheck(['GET'], $request);
+        if (!$accessCheck['status']) {
+            return $accessCheck['message'];
+        }
+
+        // Power level values.
+        $powerLevelDefault = 0;
+        $powerLevelModerator = 50;
+        $powerLevelMaximum = 100;
+
+        $entityManager = $this->getDoctrine()->getManager();
+
+        // Room info.
+        $room = $entityManager->getRepository(Room::class)->findOneBy([
```

Comment:
How does it know which room to fetch? This will only fetch the first.

This request should fetch all rooms where the requesting user is a member.
In our case the requesting user is our admin user but I don't think we put the admin user in the room (we should, but separate issue to solve).

The request _may_ (does in our case) include a filter which we also need to honour.

Essentially this needs to be:

```
// Need to return a new 'Filter' object I think to help us parse the provided filter and turn it into something meaningful.
$filter = Filter::fromRequest($request);

$roomSearch = [
    'serverid' => $serverID,
];

if ($roomIds = $filter->getRoomFilter()->getRoomIds()) {
    // Doctrine will convert an array into a WHERE IN.
    $roomSearch['roomid'] = $roomIds;
}

$roomList = $entityManager->getRepository(Room::class)->findBy($roomSearch);

$roomData = [];
foreach ($roomList as $room) {
    $roomData[$room->getRoomId()] = [
        'state' => $room->getRoomState($request),
    ];
}

// ...
return new JsonResponse((object) [
    'rooms' = [
        'join' => $roomData,
    ],
]);
```

And add a new method to either the Entity:

```
namespace App\Entity;
class Room
{
    // ...
    public function getRoomState(Filter $filter): array {
        // We pass the Filter, but we don't actually use it yet.
        // We will probably need to in the future.
        // At the moment we only know how to serve a set of power levels.
        $roomData = [
            'state' => [
                'events' => [],
            ],
        ];

        $roomData['state']['events'][] = $this->getPowerLevelState();

        return $roomData;
    }

    protected function getPowerLevelState(): array {
        // Power level values.
        $powerLevelDefault = 0;
        $powerLevelModerator = 50;
        $powerLevelMaximum = 100;

        // Get the members of the room who have a powerlevel set.
        $members = $this->getMembers()->filter(function (RoomMember $member) {
            return $member->getPowerLevel() !== null;
        });

        // Build the power level state.
        // Note: Not 100% sure on this bit without testing it. This may return an array of arrays and need flattening and there may be a better way of doing it.
        $memberInfo = $members->map(fn(RoomMember $member) => [
            $member->getUser()->getUserid() => $member->getPowerLevel(),
        ])->toArray();

        return [
            'type' => 'm.room.power_levels',
            'content' => [
                'users' => (object) $memberInfo,
                'users_default' => $powerLevelDefault,
                'events' => [
                    'm.room.name' => $powerLevelModerator,
                    'm.room.power_levels' => $powerLevelMaximum,
                    'm.room.history_visibility' => $powerLevelMaximum,
                    'm.room.canonical_alias' => $powerLevelModerator,
                    'm.room.avatar' => $powerLevelModerator,
                    'm.room.tombstone' => $powerLevelMaximum,
                    'm.room.server_acl' => $powerLevelMaximum,
                    'm.room.encryption' => $powerLevelMaximum,
                ],
                'events_default' =>  $powerLevelDefault,
                'state_default' => $powerLevelModerator,
                'ban' => $powerLevelModerator,
                'kick' => $powerLevelModerator,
                'redact' => $powerLevelModerator,
                'invite' =>  $powerLevelDefault,
                'historical' => $powerLevelMaximum,
            ],
        ];
    }
}
```

And then the Filter class:
```
// Not sure of the best location for this. Maybe Service? Maybe Util?
namespace App\Service;

use App\Service\Filter\RoomFilter;
use Symfony\Component\HttpFoundation\Request;

class Filter {
    private array $roomIds = [];
    private RoomFilter $roomFilter;

    public static function fromRequest(Request $request): static
    {
        $filter = new static();

        $payload = $request->toArray();
        if (!array_key_exists('filter', $payload)) {
            return $filter;
        }

        $filterData = $payload['filter'];

        // Filter the event data according to the API:
        // https://spec.matrix.org/v1.1/client-server-api/#filtering
        // Filters are made up of different filter types.
        // Detailed here: https://spec.matrix.org/v1.1/client-server-api/#post_matrixclientv3useruseridfilter
        // We only need to implement a handful of these.
        if (array_key_exists('room', $filterData)) {
            $filter->setRoomFilter(RoomFilter::fromArray($filterData['room']));
        }

        return $filter;
    }

    public function setRoomFilter(RoomFilter $roomFilter): static
    {
        $this->roomFilter = $roomFilter;
        return $this;
    }

    public function getRequestedRoomIds(): array
    {
        return $this->roomIds;
    }

    public function getRoomFilter(): RoomFilter
    {
        if (empty($this->roomFilter)) {
            $this->roomFilter = new RoomFilter();
        }

        return $this->roomFilter;
    }
}
```

And then corresponding RoomFilter:
```
<?php

namespace App\Service\Filter;

class RoomFilter
{
    private array $roomIds = [];
    private StateFilter $roomStateFilter;

    public static function fromArray(array $filterData): static
    {
        $filter = new static();

        // Check for a list of rooms.
        if (array_key_exists('rooms', $filterData)) {
            $filter->setRoomIds($filterData['rooms']);
        }

        // Check for a room state filter.
        if (array_key_exists('state', $filterData)) {
            $filter->setRoomState(StateFilter::fromArray($filterData['state']));
        }

        return $filter;
    }

    public function setRoomIds(array $ids): static
    {
        $this->roomIds = $ids;
        return $this;
    }

    public function setRoomState(StateFilter $state): static
    {
        $this->roomStateFilter = $state;
        return $this;
    }

    public function getRoomIds(): array
    {
        return $this->roomIds;
    }

    public function getRoomState(): StateFilter
    {
        return $this->roomStateFilter;
    }
}

```

And StateFilter:
```
<?php

namespace App\Service\Filter;

class StateFilter
{
    private array $roomIds = [];
    private StateFilter $roomStateFilter;

    public static function fromArray(array $filterData): static
    {
        $filter = new static();

        // We currently only support types.
        if (array_key_exists('types', $filterData)) {
            $filter->setTypes($filterData['types']);
        }

        return $filter;
    }

    public function setTypes(array $types): static
    {
        $this->types = $types;
        return $this;
    }

    public function getTypes(): array
    {
        return $this->types;
    }
}
```

Reply:
Thank you for providing this. It really helped.

I had to make some adjustments. Nothing major.

All communication tests are passing in Behat and PHPUnit with these changes.